common/headers/versionChecker.go

```diff
@@ -52,39 +52,43 @@ var (
 		ClientNameCLI:     "<2.0.0",
 		ClientNameServer:  "<2.0.0",
 	}
+
+	// Clients that support following next execution run id for other workflow completion events.
+	NewRetryClients = map[string]string{
+		ClientNameGoSDK:   ">=1.10.0",
+		ClientNameJavaSDK: ">=1.4.0",
+		ClientNameServer:  ">=1.13.0",
+	}
 )
 
 type (
 	// VersionChecker is used to check client/server compatibility and client's capabilities
 	VersionChecker interface {
 		ClientSupported(ctx context.Context, enableClientVersionCheck bool) error
+		ClientSupportsNewRetryEvents(ctx context.Context) bool
```

Comment:
Consider adding a more generic implementation. Something like
```
ClientSupportsFeature(ctx, feature Feature) bool
```
where Feature is enum.

Reply:
Yes, the idea of supported features was to to decouple feature names and SDK versions. Server shouldn't check the version, it just should check if feature with specific name (I would vote for `string`) supported.

Reply:
But the SDKs don't send feature strings now. Are you saying you want this internal api to work in terms of names, without changing the protocol, or that you want the protocol to change now also?

Reply:
I don't think we should require SDK to send in explicit list of features supported. Instead, server can do the mapping based on version number. I think the suggested change is to change the func name of this API to ClientSupportsFeature() and takes a string as feature name. In side the impl of that func, do a switch on the feature name. For now, that feature name would only be something like "follow-next-run-id".

Reply:
> do the mapping based on version number

Version of what? How will the server be aware of version capabilities for SDKs we don't build?

Reply:
You could say, the same way we're aware of capabilities of SDKs that we do build: ask the people responsible, look at the code and change logs, etc.

I agree that having SDKs send feature flags instead of a single version number is cleaner in some ways, and extends better to clients/SDKs that we don't control, but there are some cons too: the feature list may get very long over time, and it's harder to tell when we can clean up backwards-compatibility code paths.

For the moment, I like the idea of the internal API being in terms of feature strings, and then versionChecker can decide whether to implement that based on versions (for now), or feature flags sent by clients, or a combination of both (e.g. _x_ is supported if we know that the client supports by version _or_ if the client explicitly claims to support it).

Reply:
We plan to have SDK deprecation policy. So the feature list can be trimmed with time. But for now, it looks OK relying on SDK versions only.